kanidm_proto/src/oauth2.rs

```diff
@@ -0,0 +1,87 @@
+use url::Url;
+use webauthn_rs::base64_data::Base64UrlSafeData;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub enum CodeChallengeMethod {
+    // default to plain if not requested as S256. Reject the auth?
+    // plain
+    // BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
+    S256,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct AuthorisationRequest {
+    // Must be "code". (or token, see 4.2.1)
+    pub response_type: String,
+    pub client_id: String,
+    pub state: Base64UrlSafeData,
+    // base64?
+    pub code_challenge: Base64UrlSafeData,
+    // Probably also should be an enum.
+    pub code_challenge_method: CodeChallengeMethod,
+    // Uri?
+    pub redirect_uri: Url,
+    // appears to be + seperated?
+    pub scope: String,
+}
+
+/// We ask our user to consent to this Authorisation Request with the
+/// following data.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ConsentRequest {
+    // A pretty-name of the client
+    pub client_name: String,
+    pub scopes: Vec<String>,
+    // The users displayname (?)
+    // pub display_name: String,
+    // The token we need to be given back to allow this to proceed
+    pub consent_token: String,
+}
+
+// The resource server then contacts the token endpoint with
+//
+#[derive(Serialize, Deserialize, Debug)]
+pub struct AccessTokenRequest {
+    // must be authorization_code
+    pub grant_type: String,
+    // As sent by the authorisationCode
+    pub code: String,
+    // Must be the same as the original redirect uri.
+    pub redirect_uri: Url,
+    // REQUIRED, if the client is not authenticating with the
+    //  authorization server as described in Section 3.2.1.
+    pub client_id: Option<String>,
```

Comment:
client_id actually refers to the `name` of the resource server right?

Reply:
Yeppppppppppppp because why would standards have clear names when we can have confusing and inconsistent terms instead! 

https://datatracker.ietf.org/doc/html/rfc6749#section-4.1

Reply:
This alrso has the flow diagram btw, which I could copy into something ?

Reply:
Yea that would be good, and show which step maps to which function in code.
I think probably the best is to have a "Step" enum like AuthStep. But I think the OAuth code in this PR is stateless?

Reply:
It's not quite stateless - the state is just encrypted into these tokens and sent between requests. The reason for this is that in authentication we know we are only interacting with a single user-agent (browser, cli tool etc), so in a high-available setup with a load balancer, we can rely on sticky-sessions so that the user-agent will always be routed to the same kanidm server. This lets us store the state internally in the server, and prevents replay attacks against authentication from occuring.

However, because oauth can come from many UA's (the user, the resource server etc), and we can't rely on replication to complete within a known time bound, we have to store the state in the requests, so that which-ever kanidm server in the HA setup recieves the request, can then proceed and handle the process. 

Now the obvious question then becomes "but what about replay attacks", but in this case (unlike auth):

* We already must be authenticated as a user, so a user can really only "replay attack" themself - since they are already authenticated they don't need to do a replay attack on oauth consent requests because they cuold just request new sessions anyway!
* The resource server can resupply an access token many times requesting a session for a user. While it would be annoying, it's not a problem again because the resource server *is* also authenticated, and the consent request does express an intent that the user does want to be authenticated. Finally the access token only lasts for 60 seconds which bounds the time frame of this possibility.
* Really, what would an attacker/user gain from replaying these events?

So with that in mind, that's why this is "stateless" in appearance :)